src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,284 @@
+                } else {
+                    builder.setHttpMethod("GET");
+                }
+                break;
+            case Request.Method.GET:
+                // Not necessary to set the request method because connection defaults to GET but
+                // being explicit here.
+                builder.setHttpMethod("GET");
+                break;
+            case Request.Method.DELETE:
+                builder.setHttpMethod("DELETE");
+                break;
+            case Request.Method.POST:
+                builder.setHttpMethod("POST");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.PUT:
+                builder.setHttpMethod("PUT");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.HEAD:
+                builder.setHttpMethod("HEAD");
+                break;
+            case Request.Method.OPTIONS:
+                builder.setHttpMethod("OPTIONS");
+                break;
+            case Request.Method.TRACE:
+                builder.setHttpMethod("TRACE");
+                break;
+            case Request.Method.PATCH:
+                builder.setHttpMethod("PATCH");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            default:
+                throw new IllegalStateException("Unknown method type.");
+        }
+    }
+
+    /**
+     * Sets the request headers for the UrlRequest.
+     *
+     * @param request to get the headers from
+     * @param additionalHeaders for the UrlRequest
+     * @param builder that we are adding the request headers to
+     * @throws AuthFailureError is thrown if Request#getHeaders throws ones
+     */
+    private void setRequestHeaders(
+            Request<?> request, Map<String, String> additionalHeaders, UrlRequest.Builder builder)
+            throws AuthFailureError {
+        HashMap<String, String> map = new HashMap<>();
+        map.putAll(additionalHeaders);
+        // Request.getHeaders() takes precedence over the given additional (cache) headers).
+        map.putAll(request.getHeaders());
+        for (Map.Entry<String, String> header : map.entrySet()) {
+            builder.addHeader(header.getKey(), header.getValue());
+        }
+    }
+
+    /** Sets the UploadDataProvider of the UrlRequest.Builder */
+    private void addBodyIfExists(@Nullable byte[] body, UrlRequest.Builder builder) {
+        if (body != null) {
+            UploadDataProvider dataProvider = UploadDataProviders.create(body);
+            builder.setUploadDataProvider(dataProvider, mCallbackExecutor);
+        }
+    }
+
+    /** Sets the priority of this request. */
+    private void setPriority(Request<?> request, UrlRequest.Builder builder) {
+        builder.setPriority(request.getPriority().ordinal());
+    }
+
+    private int getContentLength(UrlResponseInfo urlResponseInfo) {
```

Comment:
So looking more closely at how BasicNetwork#inputStreamToBytes works, I still don't quite think we've matched it.

What that method does is grab a 1K buffer from the pool and use it to read chunks of the InputStream into a PoolingByteArrayOutputStream created with the Content-Length as the expected size. What we're doing here is creating a direct buffer with the Content-Length (or 1K if unknown) to read chunks into an unsized PoolingByteArrayOutputStream (which defaults to 256 bytes).

It sounds like we can't use the pool for the ByteBuffer we use to read chunks of data from Cronet, but I think we should size it at 1K to match BasicNetwork. The response size should come into play for the PoolingByteArrayOutputStream we use to accumulate chunks of responses, rather than relying on the default size for that.

Reply:
I see what you mean, but we already have the HTTP response by the time inputStreamToBytes is called. Only thing I can think of is setting the byteArrayOutputStream once we have read the headers, but there's an issue there since that's within an inner class. I used an array to get around the variable needing to be final, but I think it's a bit weird.